codec-http2/src/main/java/io/netty/handler/codec/http2/DefaultHttp2FrameReader.java

```diff
@@ -229,28 +235,18 @@ private void processHeaderState(ByteBuf in) throws Http2Exception {
                 verifyUnknownFrame();
                 break;
         }
+        readingHeaders = false;
     }
 
     private void processPayloadState(ChannelHandlerContext ctx, ByteBuf in, Http2FrameListener listener)
                     throws Http2Exception {
-        if (in.readableBytes() < payloadLength) {
-            // Wait until the entire payload has been read.
-            return;
-        }
-
-        // Only process up to payloadLength bytes.
-        int payloadEndIndex = in.readerIndex() + payloadLength;
-
-        // We have consumed the data, next time we read we will be expecting to read a frame header.
-        readingHeaders = true;
-
         // Read the payload and fire the frame event to the listener.
         switch (frameType) {
             case DATA:
-                readDataFrame(ctx, in, payloadEndIndex, listener);
+                readDataFrame(ctx, in, payloadLength, listener);
```

Comment:
All of these methods still work despite passing in a value with different meaning only because the writer index now matches the payload length. Now that you've sliced the buffer really we don't need to pass in the end length as each method could derive it from the passed in slice itself.

Reply:
True, I'll update the method signatures to make this cleaner.